task/thermal/src/control.rs

```diff
@@ -0,0 +1,338 @@
+            }
+        }
+
+        // Read miscellaneous temperature data and log it to the sensors task
+        for (i, s) in self.bsp.misc_sensors().iter().enumerate() {
+            let post_result = match s.read_temp() {
+                Ok(v) => self.sensor_api.post(s.id, v.0),
+                Err(e) => {
+                    ringbuf_entry!(Trace::MiscReadFailed(i, e));
+                    self.read_failed_count =
+                        self.read_failed_count.wrapping_add(1);
+                    self.sensor_api.nodata(s.id, e.into())
+                }
+            };
+            if post_result.is_err() {
+                self.post_failed_count = self.post_failed_count.wrapping_add(1);
+            }
+        }
+
+        // Remember, positive margin means that all parts are happily below
+        // their max temperature; negative means someone is overheating.
+        let mut worst_margin = None;
+        let mut last_err = Ok(());
+        let power_mode = self.bsp.power_mode();
+        for (i, s) in self.bsp.inputs().iter().enumerate() {
+            let post_result = match s.sensor.read_temp() {
+                Ok(v) => {
+                    if (s.power_mode_mask & power_mode) != 0 {
+                        let margin = s.max_temp.0 - v.0;
+                        worst_margin = Some(match worst_margin {
+                            Some(m) => margin.min(m),
+                            None => margin,
+                        });
+                    }
+                    self.sensor_api.post(s.sensor.id, v.0)
+                }
+                Err(e) => {
+                    // Ignore errors if
+                    // a) this sensor shouldn't be on in this power mode, or
+                    // b) the sensor is removable and not present
+                    if (s.power_mode_mask & power_mode) != 0
+                        && !(s.removable && e == ResponseCode::NoDevice)
+                    {
+                        last_err = Err(e);
+                        ringbuf_entry!(Trace::SensorReadFailed(i, e));
+                    }
+                    self.sensor_api.nodata(s.sensor.id, e.into())
+                }
+            };
+            if post_result.is_err() {
+                self.post_failed_count = self.post_failed_count.wrapping_add(1);
+            }
+        }
+
+        // Prioritize returning errors, because they indicate that something is
+        // wrong with the sensors that are critical to the control loop.
+        last_err?;
+
+        Ok(worst_margin)
+    }
+
+    /// An extremely simple thermal control loop.
+    ///
+    /// Returns an error if the control loop failed to read critical sensors;
+    /// the caller should set us to some kind of fail-safe mode if this
+    /// occurs.
+    pub fn run_control(&mut self) -> Result<(), ThermalError> {
+        // Ignore errors and missing readings
+        // TODO: handle this better
```

Comment:
Are these being ignored? It looks like they are being propagated, leaving it to someone else to ignore them. ;)

Reply:
Yup, removed comment